lib/reports/shared_print_newly_ingested.rb

```diff
@@ -0,0 +1,172 @@
+    # which calls matching_clusters,
+    # which calls matching_item_ids,
+    # which calls matching_item_ids_from_(db|file),
+    # which gets the most basic data. Add data as we go back up the call stack.
+    def run
+      File.open(outf, "w") do |f|
+        f.puts header
+        # Given ht_items and holders, format and output to file.
+        matching_ht_items do |ht_item, min_time, holders|
+          output = [
+            ht_item.billing_entity,
+            ht_item.item_id,
+            ht_item.access,
+            min_time,
+            holders
+          ]
+          f.puts output.join("\t")
+        end
+      end
+    end
+
+    # Get full path for writing
+    def outf
+      dir = Settings.sp_newly_ingested_report_path || "/tmp"
+      FileUtils.mkdir_p(dir)
+      ymd = Time.now.strftime("%Y%m%d")
+      fname = "sp_newly_ingested_report_#{ymd}.tsv"
+      File.join(dir, fname)
+    end
+
+    # Get header
+    def header
+      [
+        "contributor",
+        "ht_id",
+        "rights_status",
+        "ingest_date",
+        "holding_orgs"
+      ].join("\t")
+    end
+
+    # Given clusters, get their items and holders
+    def matching_ht_items
+      return enum_for(:matching_ht_items) unless block_given?
+      matching_clusters do |cluster, min_time|
+        cluster.ht_items.each do |ht_item|
+          # If single part monograph
+          if ht_item.bib_fmt == "BK" && ht_item.enum_chron.empty?
+            # ... fetch holders and yield
+            holders = holders_minus_contributor(cluster, ht_item.billing_entity)
+            yield ht_item, min_time, holders
+          end
+        end
+      end
+    end
+
+    # Get all the orgs with holdings on a cluster, minus the declared contributor
+    def holders_minus_contributor(cluster, contributor)
+      cluster
+        .holdings
+        .map(&:organization)
+        .reject { |org| org == contributor }
+        .sort
+        .uniq
+        .join(",")
+    end
+
+    # Given item_ids, find the corresponding clusters
+    def matching_clusters
+      return enum_for(:matching_clusters) unless block_given?
+      matching_item_ids do |ht_id, min_time|
+        Cluster.where("ht_items.item_id": ht_id).each do |cluster|
+          yield cluster, min_time
+        end
+      end
+    end
+
+    # Yields pairs of ht_id + min_time from mysql or file
+    def matching_item_ids
+      return enum_for(:matching_item_ids) unless block_given?
+      # Ask item_ids_src which method to call
+      send(item_ids_src) do |ht_id, min_time|
+        yield ht_id, min_time
+      end
+    end
+
+    # Which method to call for item_ids?
+    def item_ids_src
+      if @ht_item_ids_file.nil?
+        :matching_item_ids_from_db
+      else
+        :matching_item_ids_from_file
+      end
+    end
+
+    # Get item_id and "min_time" from mysql, these are the volumes that
+    # were ingested or got their rights assigned since @start_date.
+    def matching_item_ids_from_db
```

Comment:
Perhaps worth nothing that this method introduces a new dependency on the rights database and also is not tested (because instead of reading from the DB you can also just get this from a file). I _think_ this is fine to leave for now, but we may want to add a comment that it should be a target for getting via an API or something in the future...

Reply:
It's not a brand new dependency, as in we already have production code that uses the rights db in similar ways. But yes, I'll put a sentence there to that effect.

Reply:
\<sniffle\> Everyone forgets poor old https://github.com/hathitrust/rights_database

(Probably not worth trying to shoehorn it in now, but worth a TODO)